q2_vsearch/plugin_setup.py

```diff
@@ -427,4 +429,46 @@
                  'documentation.')
 )
 
+plugin.visualizers.register_function(
+    function=q2_vsearch._stats.fastq_stats_paired,
+    inputs={
+        'sequences': SampleData[PairedEndSequencesWithQuality],
+    },
+    parameters={
+        'threads': qiime2.plugin.Int % qiime2.plugin.Range(
```

Comment:
Instead of making 0 a magic value, how about we do something like this, instead?

```suggestion
        'threads': qiime2.plugin.Int % qiime2.plugin.Range(1, None) | qiime2.plugin.Str % qiime2.plugin.Choices(['auto'])
```

Reply:
I get errors with this code on version 2019.4 while doing `qiime dev refresh-cache`.

If we change it, should other code in this file be changed for multi threading too? I was using the style of the other functions.

```
Traceback (most recent call last):
  File "/homes/shurka/miniconda3/envs/qiime2-2019.4.dev/bin/qiime", line 11, in <module>
    sys.exit(qiime())
  File "/homes/shurka/miniconda3/envs/qiime2-2019.4.dev/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/homes/shurka/miniconda3/envs/qiime2-2019.4.dev/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/homes/shurka/miniconda3/envs/qiime2-2019.4.dev/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/homes/shurka/miniconda3/envs/qiime2-2019.4.dev/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/homes/shurka/miniconda3/envs/qiime2-2019.4.dev/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/homes/shurka/miniconda3/envs/qiime2-2019.4.dev/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/homes/shurka/miniconda3/envs/qiime2-2019.4.dev/lib/python3.6/site-packages/q2cli/builtin/dev.py", line 31, in refresh_cache
    q2cli.core.cache.CACHE.refresh()
  File "/homes/shurka/miniconda3/envs/qiime2-2019.4.dev/lib/python3.6/site-packages/q2cli/core/cache.py", line 76, in refresh
    self._state = self._get_cached_state(refresh=True)
  File "/homes/shurka/miniconda3/envs/qiime2-2019.4.dev/lib/python3.6/site-packages/q2cli/core/cache.py", line 103, in _get_cached_state
    self._cache_current_state(current_requirements)
  File "/homes/shurka/miniconda3/envs/qiime2-2019.4.dev/lib/python3.6/site-packages/q2cli/core/cache.py", line 200, in _cache_current_state
    state = self._get_current_state()
  File "/homes/shurka/miniconda3/envs/qiime2-2019.4.dev/lib/python3.6/site-packages/q2cli/core/cache.py", line 240, in _get_current_state
    state['plugins'][name] = self._get_plugin_state(plugin)
  File "/homes/shurka/miniconda3/envs/qiime2-2019.4.dev/lib/python3.6/site-packages/q2cli/core/cache.py", line 262, in _get_plugin_state
    state['actions'][id] = self._get_action_state(action)
  File "/homes/shurka/miniconda3/envs/qiime2-2019.4.dev/lib/python3.6/site-packages/q2cli/core/cache.py", line 279, in _get_action_state
    data = {'name': name, 'repr': self._get_type_repr(spec.qiime_type),
  File "/homes/shurka/miniconda3/envs/qiime2-2019.4.dev/lib/python3.6/site-packages/q2cli/core/cache.py", line 346, in _get_type_repr
    if style.expr.predicate is not None:
  File "/homes/shurka/miniconda3/envs/qiime2-2019.4.dev/lib/python3.6/site-packages/qiime2/core/type/grammar.py", line 58, in __getattr__
    % (type(self), name))
AttributeError: <class 'qiime2.core.type.grammar.UnionExp'> object has no attribute 'predicate'
```

Reply:
Works with 2019.7. Changed only my part because there is some more work to be done in the code.